src/Polysemy/Internal/WeaveClass.hs

```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving, QuantifiedConstraints, TupleSections #-}
+{-# OPTIONS_HADDOCK not-home #-}
+module Polysemy.Internal.WeaveClass
+  ( MonadTransControl(..)
+  , controlT
+
+  , mkInitState
+  , mkDistrib
+  , Distrib(..)
+  , mkInspector
+
+  , ComposeT(..)
+  ) where
+
+import Control.Monad
+import Data.Coerce
+import Data.Functor.Identity
+import Data.Functor.Compose
+import Data.Tuple
+import Control.Monad.Trans
+import Control.Monad.Trans.Identity
+import Control.Monad.Trans.Maybe
+
+import qualified Control.Monad.Trans.Except as E
+import qualified Control.Monad.Trans.State.Lazy as LSt
+import qualified Control.Monad.Trans.State.Strict as SSt
+import qualified Control.Monad.Trans.Writer.Lazy as LWr
+
+-- | A variant of the classic @MonadTransControl@ class from @monad-control@,
+-- but with a small number of changes to make it more suitable with Polysemy's
+-- internals.
+class ( MonadTrans t
+      , forall z. Monad z => Monad (t z)
+      , Traversable (StT t)
+      )
+   => MonadTransControl t where
+  type StT t :: * -> *
+
+  hoistT :: (Monad m, Monad n)
+         => (forall x. m x -> n x)
+         -> t m a -> t n a
+  hoistT n m = controlT $ \lower -> n (lower m)
+  {-# INLINE hoistT #-}
+
+  liftWith :: Monad m
+           => ((forall z x. Monad z => t z x -> z (StT t x)) -> m a)
+           -> t m a
+
+  restoreT :: Monad m => m (StT t a) -> t m a
+
+controlT :: (MonadTransControl t, Monad m)
+         => ((forall z x. Monad z => t z x -> z (StT t x)) -> m (StT t a))
+         -> t m a
+controlT main = liftWith main >>= restoreT . pure
+{-# INLINE controlT #-}
+
+newtype ComposeT t (u :: (* -> *) -> * -> *) m a = ComposeT {
+    getComposeT :: t (u m) a
+  }
+  deriving (Functor, Applicative, Monad)
+
+instance ( MonadTrans t
+         , MonadTrans u
+         , forall m. Monad m => Monad (u m)
+         )
+      => MonadTrans (ComposeT t u) where
+  lift m = ComposeT (lift (lift m))
+
+instance ( MonadTransControl t
+         , MonadTransControl u
+         )
+      => MonadTransControl (ComposeT t u) where
+  type StT (ComposeT t u) = Compose (StT u) (StT t)
+
+  hoistT n (ComposeT m) = ComposeT (hoistT (hoistT n) m)
+
+  liftWith main = ComposeT $
+    liftWith $ \lowerT ->
+    liftWith $ \lowerU ->
+    main (\(ComposeT m) -> Compose <$> lowerU (lowerT m))
+
+  restoreT m = ComposeT (restoreT (restoreT (fmap getCompose m)))
+
+newtype Distrib f q m = Distrib (forall x. f (q x) -> m (f x))
```

Comment:
Hack: lets you bind the distribution function in a `let` without the monomorphism restriction kicking in.

Reply:
filthy little monolocalbindses!